opm/simulators/utils/ParallelEclipseState.cpp

```diff
@@ -0,0 +1,310 @@
+           serializer.packSize(m_runspec) +
+           serializer.packSize(m_eclipseConfig) +
+           serializer.packSize(m_deckUnitSystem) +
+           serializer.packSize(m_inputNnc) +
+           serializer.packSize(m_inputEditNnc) +
+           serializer.packSize(m_simulationConfig) +
+           serializer.packSize(m_transMult) +
+           serializer.packSize(m_faults) +
+           serializer.packSize(m_title);
+
+}
+
+
+void ParallelEclipseState::pack(std::vector<char>& buffer, int& position,
+                                EclMpiSerializer& serializer) const
+{
+    serializer.pack(m_tables, buffer, position);
+    serializer.pack(m_runspec, buffer, position);
+    serializer.pack(m_eclipseConfig, buffer, position);
+    serializer.pack(m_deckUnitSystem, buffer, position);
+    serializer.pack(m_inputNnc, buffer, position);
+    serializer.pack(m_inputEditNnc, buffer, position);
+    serializer.pack(m_simulationConfig, buffer, position);
+    serializer.pack(m_transMult, buffer, position);
+    serializer.pack(m_faults, buffer, position);
+    serializer.pack(m_title, buffer, position);
+}
+
+
+void ParallelEclipseState::unpack(std::vector<char>& buffer, int& position,
+                                  EclMpiSerializer& serializer)
+{
+    serializer.unpack(m_tables, buffer, position);
+    serializer.unpack(m_runspec, buffer, position);
+    serializer.unpack(m_eclipseConfig, buffer, position);
+    serializer.unpack(m_deckUnitSystem, buffer, position);
+    serializer.unpack(m_inputNnc, buffer, position);
+    serializer.unpack(m_inputEditNnc, buffer, position);
+    serializer.unpack(m_simulationConfig, buffer, position);
+    serializer.unpack(m_transMult, buffer, position);
+    serializer.unpack(m_faults, buffer, position);
+    serializer.unpack(m_title, buffer, position);
+}
+
+
+const FieldPropsManager& ParallelEclipseState::fieldProps() const
+{
+    if (!m_parProps)
+        assert(Dune::MPIHelper::getCollectiveCommunication().rank() == 0);
+
+    if (!m_parProps || Dune::MPIHelper::getCollectiveCommunication().size() == 1)
+        return this->EclipseState::fieldProps();
+
+    return m_fieldProps;
+}
+
+
+const FieldPropsManager& ParallelEclipseState::globalFieldProps() const
+{
+    assert(Dune::MPIHelper::getCollectiveCommunication().rank() == 0);
+    return this->EclipseState::globalFieldProps();
+}
+
+
+const EclipseGrid& ParallelEclipseState::getInputGrid() const
+{
+    assert(Dune::MPIHelper::getCollectiveCommunication().rank() == 0);
+    return this->EclipseState::getInputGrid();
+}
+
+
+void ParallelEclipseState::switchToGlobalProps()
+{
+    m_parProps = false;
+}
+
+
+void ParallelEclipseState::switchToDistributedProps()
+{
+    const auto& comm = Dune::MPIHelper::getCollectiveCommunication();
+    if (comm.size() == 1) // No need for the parallel frontend
+        return;
+
+    m_parProps = true;
+}
+
+void ParallelEclipseState::setupLocalProps(const std::vector<int>& localToGlobal)
```

Comment:
Took me quite some time to understand this method (maybe a short comment would have helped).
I gues, this is supposed to be an intermediate solution?

Not 100% sure when the fieldprops are available, but if they are during loadbalance, then this might have been solvable with communication datahandle (`CommDataHandleIF`) that gets passed to it. But one would need to send the integers as double as the handle only supports one type. For `CpGrid` there is also `scatterData`.

In addition there is also an implementation of this in `ebos/collecttoiorank,hh``called `DistributedIndexMapping`. I am not saying that one is better than the other, but having two is definitely more maintenance and adds to the confusion.

Reply:
yes, this is intermediate. i am still waiting for a non-cpgrid build before i attempt doing it in a general fashion. it can certainly be done as you suggest.

Reply:
Me, too (if by build, you mean test). Maybe the developers from Norce can provida an ALUGrid test that does not need dune-fem?

Reply:
Postponed and resolved.